src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTests.cs

```diff
@@ -2,272 +2,180 @@
+                throw new InvalidOperationException($"Type {original} has no base64 hashes to deserialize and test equality against. " +
+                    $"Hash for object is: " + SerializeObjectToHash(original));
+            }
+
+            foreach (string tfmBase64Hash in tfmBase64Hashes)
+            {
+                CheckForAnyEquals(original, DeserializeObjectHash(original, tfmBase64Hash));
+            }
+        }
+
+        private void CheckForAnyEquals(object original, object clone)
+        {
+            if (original != null && clone != null)
+            {
+                object result = null;
+                Type originalType = original.GetType();
+
+                // Check if custom equality extension method is available
+                MethodInfo customEqualityCheck = GetExtensionMethod(typeof(EqualityExtensions).Assembly, originalType);
+                if (customEqualityCheck != null)
                 {
-                    foreach (TypeFilterLevel filterLevel in new[] { TypeFilterLevel.Full, TypeFilterLevel.Low })
+                    result = customEqualityCheck.Invoke(original, new object[] { original, clone });
+                }
+                else
+                {
+                    // Check if object.Equals(object) is overridden and if not check if there is a more concrete equality check implementation
+                    bool equalsNotOverridden = originalType.GetMethod("Equals", new Type[] { typeof(object) }).DeclaringType == typeof(object);
+                    if (equalsNotOverridden)
                     {
-                        foreach (FormatterTypeStyle typeFormat in new[] { FormatterTypeStyle.TypesAlways, FormatterTypeStyle.TypesWhenNeeded, FormatterTypeStyle.XsdString })
-                        {  
-                            yield return new object[] { obj, assemblyFormat, filterLevel, typeFormat};
+                        // If type doesn't override Equals(object) method then check if there is a more concrete implementation
+                        // e.g. if type implements IEquatable<T>.
+                        MethodInfo equalsMethod = originalType.GetMethod("Equals", new Type[] { originalType });
+                        if (equalsMethod.DeclaringType != typeof(object))
+                        {
+                            result = equalsMethod.Invoke(original, new object[] { clone });
                         }
                     }
                 }
+
+                if (result != null)
+                {
+                    Assert.True((bool)result, "Error during equality check of type " + originalType.FullName);
+                    return;
+                }
+            }
+
+            try { Assert.Equal(original, clone); }
+            catch (Exception)
+            {
+                Console.WriteLine("Error during equality check of type " + original?.GetType()?.FullName);
```

Comment:
Generally I think we try to include messages in the exception so as not to clutter console, but it's not a big deal worht changing now.